File: backend/app/schemas/shield.py
# SPDX-License-Identifier: Apache-2.0
# Copyright 2026 Mutlu Erol
"""Shield (sertifika/TLS) şemaları."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ============================================================================
# Sertifika
# ============================================================================
class CertificatePublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    type: str  # SERVER | CA | CLIENT
    serial_number: str
    subject: str
    issuer: str
    not_before: datetime
    not_after: datetime
    fingerprint_sha256: str
    is_active: bool
    has_private_key: bool
    description: str | None = None
    source: str
    activated_at: datetime | None = None
    uploaded_at: datetime
    # türetilmiş alanlar (servis dolduruyor)
    days_remaining: int | None = None
    is_expired: bool | None = None
    is_self_signed: bool | None = None


class CertificateDetail(CertificatePublic):
    """Tek sertifika detayı — PEM dahil."""

    pem_data: str


class CertificateUploadRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    type: Literal["SERVER", "CA", "CLIENT"]
    pem: str = Field(..., min_length=64, description="Public sertifika PEM")
    private_key: str | None = Field(
        None, description="SERVER tipi için özel anahtar PEM (diskte 600 saklanır, DB'ye girmez)"
    )
    csr_id: UUID | None = Field(
        None,
        description="Panel üretimli CSR'ı karşılarken: anahtar diskteki CSR kaydından alınır (private_key gerekmez)",
    )
    description: str | None = Field(None, max_length=512)


class CertificateActivateResponse(BaseModel):
    certificate: CertificatePublic
    slapd_reloaded: bool
    nginx_reloaded: bool
    ca_trust_updated: bool
    live_ldaps_fingerprint: str | None = None
    live_https_fingerprint: str | None = None
    message: str
    # CA değişiminde replikasyon uyarısı
    replication_warning: str | None = None


# ============================================================================
# Genel bakış (overview)
# ============================================================================
class LiveEndpointStatus(BaseModel):
    name: str  # "ldaps" | "https"
    host: str
    port: int
    reachable: bool
    fingerprint_sha256: str | None = None
    matches_active: bool | None = None  # canlı cert == aktif SERVER cert mi?
    error: str | None = None


class ShieldOverview(BaseModel):
    active_ca: CertificatePublic | None = None
    active_server: CertificatePublic | None = None
    total_certificates: int
    pending_csr: int
    endpoints: list[LiveEndpointStatus]
    # uyarılar
    warnings: list[str] = Field(default_factory=list)


# ============================================================================
# CSR
# ============================================================================
class CsrGenerateRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    common_name: str = Field(..., min_length=1, max_length=255, description="CN — örn. mtl-master-01.mtl.local")
    organization: str = Field("MTL", max_length=128)
    country: str = Field("TR", min_length=2, max_length=2)
    san_dns: list[str] = Field(default_factory=list, description="Ek DNS SAN'ları")
    san_ip: list[str] = Field(default_factory=list, description="Ek IP SAN'ları")
    key_bits: Literal[2048, 4096] = 4096


class CsrPublic(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    name: str
    subject: str
    status: str
    key_fingerprint: str | None = None
    fulfilled_cert_id: UUID | None = None
    created_at: datetime
    fulfilled_at: datetime | None = None


class CsrGenerateResponse(BaseModel):
    csr: CsrPublic
    csr_pem: str  # indirilebilir CSR metni


# ============================================================================
# CA geçişi / dışa aktarım
# ============================================================================
class CaExportResponse(BaseModel):
    name: str
    pem: str
    fingerprint_sha256: str
    subject: str
    not_after: datetime
    # slave dağıtımı için bilgi
    note: str


class CaResignRequest(BaseModel):
    """Mevcut bir CSR'ı VEYA serbest CN'i MTL CA ile imzala (geçiş/test için)."""

    csr_id: UUID | None = Field(None, description="Panel üretimli CSR id (varsa)")
    name: str = Field(..., min_length=1, max_length=128)
    description: str | None = Field(None, max_length=512)
    days: int = Field(1825, ge=1, le=3650)


class TransitionActivateRequest(BaseModel):
    """CA geçişi: yeni CA + yeni server sertifikasını ATOMİK aktive et.

    server cert, ca'ya zincirlenmeli; helper bunu doğrular, başarısızsa rollback yapar.
    """

    ca_id: UUID = Field(..., description="Aktive edilecek CA sertifikası (envanter id)")
    server_id: UUID = Field(..., description="Aktive edilecek SERVER sertifikası (envanter id)")

File: frontend/src/types/shield.ts
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 Mutlu Erol
//
// Backend: app/schemas/shield.py

export type CertType = "SERVER" | "CA" | "CLIENT";
export type CertSource = "UPLOAD" | "GENERATED" | "INSTALLER" | "RESIGNED";

export interface Certificate {
  id: string;
  name: string;
  type: string; // CertType
  serial_number: string;
  subject: string;
  issuer: string;
  not_before: string;
  not_after: string;
  fingerprint_sha256: string;
  is_active: boolean;
  has_private_key: boolean;
  description: string | null;
  source: string; // CertSource
  activated_at: string | null;
  uploaded_at: string;
  days_remaining: number | null;
  is_expired: boolean | null;
  is_self_signed: boolean | null;
}

export interface CertificateDetail extends Certificate {
  pem_data: string;
}

export interface CertUploadPayload {
  name: string;
  type: CertType;
  pem: string;
  private_key?: string | null;
  csr_id?: string | null;
  description?: string | null;
}

export interface CertActivateResponse {
  certificate: Certificate;
  slapd_reloaded: boolean;
  nginx_reloaded: boolean;
  ca_trust_updated: boolean;
  live_ldaps_fingerprint: string | null;
  live_https_fingerprint: string | null;
  message: string;
  replication_warning: string | null;
}

export interface LiveEndpointStatus {
  name: string; // "ldaps" | "https"
  host: string;
  port: number;
  reachable: boolean;
  fingerprint_sha256: string | null;
  matches_active: boolean | null;
  error: string | null;
}

export interface ShieldOverview {
  active_ca: Certificate | null;
  active_server: Certificate | null;
  total_certificates: number;
  pending_csr: number;
  endpoints: LiveEndpointStatus[];
  warnings: string[];
}

export interface Csr {
  id: string;
  name: string;
  subject: string;
  status: string; // PENDING | FULFILLED | CANCELLED
  key_fingerprint: string | null;
  fulfilled_cert_id: string | null;
  created_at: string;
  fulfilled_at: string | null;
}

export interface CsrGeneratePayload {
  name: string;
  common_name: string;
  organization: string;
  country: string;
  san_dns: string[];
  san_ip: string[];
  key_bits: 2048 | 4096;
}

export interface CsrGenerateResponse {
  csr: Csr;
  csr_pem: string;
}

export interface CaExportResponse {
  name: string;
  pem: string;
  fingerprint_sha256: string;
  subject: string;
  not_after: string;
  note: string;
}

export interface CaResignPayload {
  name: string;
  description?: string | null;
  days: number;
}

export interface TransitionActivatePayload {
  ca_id: string;
  server_id: string;
}
